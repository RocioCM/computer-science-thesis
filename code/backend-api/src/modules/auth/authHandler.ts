import { StatusCodes } from 'http-status-codes';
import { IResult } from 'src/pkg/interfaces/result';
import { CreateUserDTO, UpdateUserDTO, User } from './domain/user';
import FirebaseAuthRepository from './repositories/firebaseAuthRepository';
import UserRepository from './repositories/userRepository';

export default class AuthHandler {
  static async RegisterUser(user: CreateUserDTO): IResult<User> {
    const {
      ok,
      status,
      data: firebaseUser,
    } = await FirebaseAuthRepository.CreateUser(user.email, user.password);
    if (!ok) {
      return { ok: false, status, data: null };
    }
    await FirebaseAuthRepository.SetUserRole(firebaseUser.uid, user.roleId);
    const { ok: ok2, data: blockchainId } =
      await UserRepository.GetAvailableBlockchainId();
    if (!ok2) {
      return {
        ok: false,
        status: StatusCodes.INTERNAL_SERVER_ERROR,
        data: null,
      };
    }

    const userEntity = new User();
    userEntity.firebaseUid = firebaseUser.uid;
    userEntity.email = user.email;
    userEntity.blockchainId = blockchainId;

    return UserRepository.CreateUser(userEntity);
  }

  static async GetUserWithAuth(firebaseUid: string): IResult<User> {
    const userRes = await UserRepository.GetUserByFirebaseUid(firebaseUid);
    if (!userRes.ok) {
      return { ok: false, status: StatusCodes.NOT_FOUND, data: null };
    }

    return { ok: true, status: StatusCodes.OK, data: userRes.data };
  }

  static async GetUserByFirebaseUid(firebaseUid: string): IResult<User> {
    return UserRepository.GetUserByFirebaseUid(firebaseUid);
  }

  static async GetUserByBlockchainId(address: string): IResult<User> {
    return UserRepository.GetUserByBlockchainId(address);
  }

  static async UpdateUserWithAuth(
    firebaseUid: string,
    user: UpdateUserDTO,
  ): IResult<null> {
    const userRes = await UserRepository.GetUserByFirebaseUid(firebaseUid);
    if (!userRes.ok) {
      return { ok: false, status: StatusCodes.NOT_FOUND, data: null };
    }

    const updatedUser = { ...userRes.data, ...user };

    return UserRepository.UpdateUser(updatedUser);
  }

  static async GetFilteredUsers(
    searchQuery: string,
    role?: number,
  ): IResult<User[]> {
    const matchedUsersRes = await UserRepository.GetFilteredUsers(searchQuery);
    if (!matchedUsersRes.ok) {
      return {
        ok: false,
        status: StatusCodes.INTERNAL_SERVER_ERROR,
        data: null,
      };
    }

    let filteredUsers: User[] = [];

    if (role) {
      // If role is provided, filter users by role as well.
      // Loop over the matched users and get the user from Firebase to check the role.
      // We can loop fearlessly here because the number of users in the list will be at most 10.
      for (const user of matchedUsersRes.data) {
        const userRes = await FirebaseAuthRepository.GetUserById(
          user.firebaseUid,
        );
        if (userRes.ok && userRes.data.customClaims?.role === role) {
          filteredUsers.push(user);
        }
      }
    } else {
      filteredUsers = matchedUsersRes.data;
    }

    return { ok: true, status: StatusCodes.OK, data: filteredUsers };
  }
}
