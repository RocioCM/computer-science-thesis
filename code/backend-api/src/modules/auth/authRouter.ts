import { Router, Request, Response } from 'express';
import requestHelper from 'src/pkg/helpers/requestHelper';
import responseHelper from 'src/pkg/helpers/responseHelper';
import AuthHandler from './authHandler';
import middlewareHelper from 'src/pkg/helpers/middlewareHelper';
import { CreateUserDTO, UpdateUserDTO } from './domain/user';
import { Authenticate } from 'src/pkg/helpers/authHelper';

//---- Routers ----//

async function RegisterUser(req: Request, res: Response) {
  const parsedUserRes = await requestHelper.parseBody(req.body, CreateUserDTO);
  if (!parsedUserRes.ok) {
    responseHelper.build(res, parsedUserRes.status, parsedUserRes.data);
    return;
  }

  const { status, data } = await AuthHandler.RegisterUser(parsedUserRes.data);

  responseHelper.build(res, status, data);
}

async function GetUserWithAuth(req: Request, res: Response) {
  const userRes = await Authenticate(req);
  if (!userRes.ok) {
    responseHelper.build(res, userRes.status, userRes.data);
    return;
  }

  const { status, data } = await AuthHandler.GetUserWithAuth(userRes.data.uid);

  responseHelper.build(res, status, data);
}

async function UpdateUserWithAuth(req: Request, res: Response) {
  const userRes = await Authenticate(req);
  if (!userRes.ok) {
    responseHelper.build(res, userRes.status, userRes.data);
    return;
  }

  const parsedUserRes = await requestHelper.parseBody(req.body, UpdateUserDTO);
  if (!parsedUserRes.ok) {
    responseHelper.build(res, parsedUserRes.status, parsedUserRes.data);
    return;
  }

  const { status, data } = await AuthHandler.UpdateUserWithAuth(
    userRes.data.uid,
    parsedUserRes.data,
  );

  responseHelper.build(res, status, data);
}

//---- Routes configuration ----//

const AuthRouter = Router();

middlewareHelper.applyAsyncHandlerMiddleware(AuthRouter);

AuthRouter.post('/register', RegisterUser);
AuthRouter.get('/user', GetUserWithAuth);
AuthRouter.put('/user', UpdateUserWithAuth);

export default AuthRouter;
