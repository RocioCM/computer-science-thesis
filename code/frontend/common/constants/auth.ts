export const ROLES = {
  1: 'admin',
  2: 'producer',
  3: 'secondary_producer',
  4: 'consumer',
  5: 'recycler',
  6: 'buyer',
  admin: 1,
  producer: 2,
  secondary_producer: 3,
  consumer: 4,
  recycler: 5,
  buyer: 6,
};

export const ROLES_OPTIONS = [
  { label: 'Administrador', value: ROLES.admin },
  { label: 'Productor primario', value: ROLES.producer },
  { label: 'Productor secundario', value: ROLES.secondary_producer },
  { label: 'Comprador', value: ROLES.buyer },
  { label: 'Consumidor', value: ROLES.consumer },
  { label: 'Reciclador', value: ROLES.recycler },
];
