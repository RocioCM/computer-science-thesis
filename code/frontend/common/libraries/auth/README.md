# Auth Library

This library is used to manage the authentication of the application. It provides a set of hooks and components to manage the authentication flow.
