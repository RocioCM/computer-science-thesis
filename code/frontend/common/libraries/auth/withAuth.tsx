import { useEffect, useMemo } from 'react';
import { NextComponentType, NextPageContext } from 'next';
import { useRouter } from 'next/router';
import { useLoginContext } from '.';
import { UserRole } from './types';

/**
 * Validates user is authenticated and has permission to access the requested page.
 *
 * Renders the page if the user is allowed to access it.
 *
 * Stays in an eternal loading screen if user is not authenticated (login must be done on Flutter app).
 *
 * Redirects to Error page if user is authenticated but is not allowed to access the page.
 *
 * @param Component Page to be rendered if access is allowed.
 * @param roles Array of allowed roles. All roles are allowed if not present.
 */
const withAuth = (
  Component: NextComponentType<NextPageContext, any, any>,
  roles: UserRole | UserRole[] = []
) =>
  function AuthWrapper(props: any) {
    const { loading, isLoggedIn, user, userHasRole } = useLoginContext();
    const router = useRouter();

    /** True if user is logged in. Guarantees consistency with user object. */
    const isAuthenticated = isLoggedIn && !!user;

    /** True if logged in user is authorized to access the page. */
    const isAuthorized = useMemo(
      () =>
        isLoggedIn &&
        (Array.isArray(roles)
          ? !roles.length || roles.some((r) => userHasRole(r))
          : userHasRole(roles)),
      [isLoggedIn, user, roles]
    );

    useEffect(() => {
      if (!loading) {
        if (!isAuthenticated) {
          router.replace('/auth/login'); // User not allowed.
        }
      }
    }, [loading]);

    if (loading || !isAuthenticated) return null; // User not logged in.
    if (!isAuthorized) {
      return (
        <div className="flex flex-col items-center justify-center h-full w-full">
          <h2 className="text-center">
            No tienes permiso <br /> para acceder a esta p√°gina
          </h2>
        </div>
      ); // User not allowed.
    }
    return <Component {...props} />; // User logged in and authorized. */
  };

export default withAuth;
