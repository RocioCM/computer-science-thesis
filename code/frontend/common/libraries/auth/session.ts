import { isServerSide } from '@/common/constants';

const SESSION_KEY = 'sessionToken';

// Functions to manage the session throughout the entire application.

export const getSession = (): string | null => {
  if (isServerSide) return null;
  return localStorage.getItem(SESSION_KEY);
};

export const updateSession = (session: string): void => {
  if (isServerSide) return;
  localStorage.setItem(SESSION_KEY, session);
};

export const removeSession = (): void => {
  if (isServerSide) return;
  localStorage.removeItem(SESSION_KEY);
};
