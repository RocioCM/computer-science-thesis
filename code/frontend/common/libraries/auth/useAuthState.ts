import { useReducer, useEffect } from 'react';
import { useRouter } from 'next/router';
import { reducer, initialState } from './reducer';
import AuthServices, { AuthUserInfo } from './services/FirebaseAuthServices';
import UserServices from './services/UserServices';
import {
  User,
  LoginPayload,
  ActionTypes,
  UserRole,
  RegisterUser,
  APIUser,
} from './types';
import { updateSession } from './session';

/**
 * Returns state and methods related to user auth.
 * @WARN ONLY USE this hook in a CONTEXT PROVIDER.
 * Otherwise you won't be able to share this state between screens or components.
 */
const useAuthState = () => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const router = useRouter();

  /** Joins the Auth User Info object with the User Data object from the API
   * to a friendlier structure to save in the context. */
  const formatUser = async (authUser: AuthUserInfo) => {
    const { ok, data: apiUser } = await UserServices.getUserData();
    if (!ok) return { ok: false, data: null };

    const user: User = {
      role: authUser.role,
      sessionToken: authUser.token,
      ...apiUser,
    };
    return { ok: true, data: user };
  };

  /**
   * Receives Auth User Info and saves it in the context
   *  including the User Data from the API.
   * @param authData - Auth User Info object.
   * @returns ok is true if the user was saved successfully.
   */
  const loadUserInContext = async (authData: AuthUserInfo) => {
    updateSession(authData.token);
    const { ok, data: user } = await formatUser(authData);
    if (ok && user) {
      dispatch({ type: ActionTypes.LOGIN, value: user });
      return { ok: true, data: user };
    }
    return { ok: false, data: null };
  };

  /** Refreshes user information using the available sessionToken.
   * @returns ok is true if the user data was refreshed in the context successfully.
   */
  const refreshUser = async () => {
    const { ok, data } = await AuthServices.refreshUser();
    if (ok) return loadUserInContext(data);
    return { ok: false, data: null };
  };

  const registerUser = async (newUser: RegisterUser) => {
    const res = await UserServices.registerUser(newUser);
    return res;
  };

  /**
   * Updates user data in the API and refreshes the user in the context.
   * @param user - User object with the new data.
   * @returns ok is true if the user was updated successfully.
   */
  const updateUserData = async (newUser: APIUser) => {
    const { ok } = await UserServices.updateUser(newUser);
    if (ok) refreshUser();
    return { ok };
  };

  /**
   * Triggers the Google login popup and saves the user in the context if successful.
   * @returns ok is true if the user was saved successfully.
   */
  const loginWithGoogle = async () => {
    const { ok, data } = await AuthServices.loginWithGoogle();
    if (ok) {
      return loadUserInContext(data);
    }
    return { ok: false, data: null };
  };

  /**
   * Logs in the user with email and password and saves it in the context if successful.
   * @param form - Object containing email and password.
   * @returns ok is true if the user was authenticated and saved successfully.
   */
  const loginWithPassword = async (form: LoginPayload) => {
    const { ok, data } = await AuthServices.loginWithPassword(
      form.email,
      form.password
    );
    if (ok) {
      return loadUserInContext(data);
    }
    return { ok: false, data: null };
  };

  /**
   * Logs out the user (invalidates current session) and removes it from the context.
   * @param triggerReload - If true, it will reload the page after logout.
   * @returns ok is true if the user was logged out successfully.
   */
  const logout = async (triggerReload = false) => {
    await AuthServices.logout();
    dispatch({ type: ActionTypes.LOGOUT });
    if (triggerReload) router.reload();
  };

  /** Returns true if current user has the provided role.
   * If no user is provided, it uses the current user in the state.
   * If no user is in the state, it returns false.
   */
  const userHasRole = (
    role: UserRole,
    validationUser: User | null = null
  ): boolean => {
    const user = validationUser || state.user;
    if (!user) return false;
    return user.role === role;
  };

  useEffect(() => {
    return AuthServices.initAuthStateListener(loadUserInContext, logout);
  }, []);

  // Re-export stateless Methods
  const resetPassword = AuthServices.resetPassword;
  const sendRecoverPasswordMail = AuthServices.sendRecoverPasswordMail;

  const actions = {
    userHasRole,
    refreshUser,
    registerUser,
    loginWithGoogle,
    loginWithPassword,
    updateUserData,
    logout,
    resetPassword,
    sendRecoverPasswordMail,
  };

  return { ...state, ...actions };
};

useAuthState.defaultReturn = (): ReturnType<typeof useAuthState> => ({
  loading: true,
  isLoggedIn: false,
  user: null,
  registerUser: async () => ({ ok: false, data: null, status: 0 }),
  refreshUser: async () => ({ ok: false, data: null }),
  loginWithGoogle: async () => ({ ok: false, data: null }),
  loginWithPassword: async () => ({ ok: false, data: null }),
  logout: async () => {},
  userHasRole: () => false,
  updateUserData: async () => ({ ok: false }),
  resetPassword: async () => ({ ok: false, data: null, status: 0 }),
  sendRecoverPasswordMail: async () => ({ ok: false, data: null, status: 0 }),
});

export default useAuthState;
