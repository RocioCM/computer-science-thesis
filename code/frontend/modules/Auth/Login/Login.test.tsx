import TestAppWrapper from '@/common/utils/tests';
import LoginPage from '@/pages/auth/login';
import { render, screen, fireEvent, act } from '@testing-library/react';
import { toast } from 'react-toastify';
import auth from '@/common/libraries/auth';

describe('Login Page', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Matches snapshot', () => {
    const { container } = render(
      <TestAppWrapper>
        <LoginPage />
      </TestAppWrapper>
    );
    expect(container).toMatchSnapshot();
  });

  it('does not submit with empty fields', async () => {
    const mockLoginWithPassword = jest.fn();
    (auth as jest.Mock).mockReturnValue({
      loginWithPassword: mockLoginWithPassword,
    });

    const toastErrorSpy = jest.spyOn(toast, 'error');

    render(
      <TestAppWrapper>
        <LoginPage />
      </TestAppWrapper>
    );

    const [emailInput, passwordInput] = screen.queryAllByTestId('input');
    const submitButton = screen.getByRole('button', {
      name: /iniciar sesión/i,
    });

    await act(async () => {
      fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
      passwordInput.enterKeyHint = 'done'; // Simulate pressing Enter key
      fireEvent.change(passwordInput, { target: { value: '' } });
      fireEvent.click(submitButton);
    });

    expect(mockLoginWithPassword).not.toHaveBeenCalled(); // Ensure loginWithPassword was not called
    expect(toastErrorSpy).not.toHaveBeenCalled(); // Ensure toast.error was not called
  });

  it('Displays error when login fails', async () => {
    const mockLoginWithPassword = jest.fn().mockResolvedValue({ ok: false });
    (auth as jest.Mock).mockReturnValue({
      loginWithPassword: mockLoginWithPassword,
    });

    render(
      <TestAppWrapper>
        <LoginPage />
      </TestAppWrapper>
    );

    const [emailInput, passwordInput] = screen.queryAllByTestId('input');
    const submitButton = screen.getByRole('button', {
      name: /iniciar sesión/i,
    });

    await act(async () => {
      fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
      fireEvent.change(passwordInput, { target: { value: 'wrongpassword' } });
      fireEvent.click(submitButton);
    });

    expect(mockLoginWithPassword).toHaveBeenCalledWith({
      email: 'test@example.com',
      password: 'wrongpassword',
    });
    expect(toast.error).toHaveBeenCalledWith(
      'Usuario o contraseña incorrectos. Intentá nuevamente.'
    );
  });

  it('Handles successful login', async () => {
    const mockLoginWithPassword = jest.fn().mockResolvedValue({ ok: true });
    (auth as jest.Mock).mockReturnValue({
      loginWithPassword: mockLoginWithPassword,
    });

    // Spy on toast.error to ensure it is not called
    const toastErrorSpy = jest.spyOn(toast, 'error');

    render(
      <TestAppWrapper>
        <LoginPage />
      </TestAppWrapper>
    );

    const [emailInput, passwordInput] = screen.queryAllByTestId('input');
    const submitButton = screen.getByRole('button', {
      name: /iniciar sesión/i,
    });

    await act(async () => {
      fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
      fireEvent.change(passwordInput, { target: { value: 'correctpassword' } });
      fireEvent.click(submitButton);
    });

    expect(mockLoginWithPassword).toHaveBeenCalledWith({
      email: 'test@example.com',
      password: 'correctpassword',
    });
    expect(toastErrorSpy).not.toHaveBeenCalled(); // Ensure toast.error was not called
  });
});
