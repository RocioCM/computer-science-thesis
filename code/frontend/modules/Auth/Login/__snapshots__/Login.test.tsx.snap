// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Matches snapshot 1`] = `
<div>
  <main
    class="w-full bg-p1-25 h-screen flex flex-row items-center justify-around gap-xl p-3xl"
  >
    <img
      alt="Iniciar sesión"
      class="absolute h-full w-full object-cover opacity-90 z-0"
      data-nimg="1"
      decoding="async"
      height="800"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=1200&q=75 1x, /_next/image?url=%2Fimg.jpg&w=3840&q=75 2x"
      style="color: transparent;"
      width="1200"
    />
    <form
      class="relative w-full min-w-[300px] max-w-[500px] h-max max-h-full overflow-auto rounded-rs border border-n2 bg-n0 shadow-e3 p-2xl flex flex-col gap-xl"
    >
      <h2
        class="font-medium"
      >
        Iniciar sesión
      </h2>
      <div
        class="flex flex-col gap-s"
      >
        <div
          class="w-full"
        >
          <span
            class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n4 font-medium"
          >
            Correo electrónico
          </span>
          <div
            class="relative flex flex-row flex-nowrap justify-between items-center w-full"
          >
            <div
              class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
            />
            <input
              class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent"
              data-testid="input"
              form="[object Object]"
              inputlabel="Correo electrónico"
              name="email"
              options=""
              placeholder=""
              required=""
              type="text"
              value=""
            />
            <div
              class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
            />
          </div>
        </div>
        <div
          class="w-full"
        >
          <span
            class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n4 font-medium"
          >
            Contraseña
          </span>
          <div
            class="relative flex flex-row flex-nowrap justify-between items-center w-full"
          >
            <div
              class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
            />
            <input
              class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent pr-10"
              data-testid="input"
              form="[object Object]"
              inputlabel="Contraseña"
              name="password"
              options=""
              placeholder=""
              required=""
              type="password"
              value=""
            />
            <div
              class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
            >
              <img
                alt="Show password"
                class="w-l h-l cursor-pointer mr-s"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                role="button"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
            </div>
          </div>
        </div>
      </div>
      <p
        class="text-center"
      >
        ¿No tienes cuenta?
         
        <a
          class="font-semibold text-p1"
          href="/auth/register"
        >
          Regístrate aquí
        </a>
      </p>
      <button
        class="relative flex items-center justify-center gap-s rounded-rs font-semibold h-10 disabled:cursor-default disabled:bg-opacity-40  focusButton bg-p1 text-n0 py-s px-m w-full"
        data-testid="button"
        disabled=""
      >
        <span>
          Ingresar
        </span>
      </button>
    </form>
  </main>
</div>
`;
